### ISSUE TITLE
Puppet provision exclude option for modules and manifests parameter

### ISSUE THREAD

**Issue by nbeernink** [mitchellh/vagrant#4280]
Can an exclude option for puppet provisioning be added to allow excludes in the puppet modules and manifests rsyncs? I have a fixtures directory used for rspec tests which contains a symlink to the modules dir its in (to reduce file duplication, reduce storage consumption):
```
$ ls -l modules/mymodule/spec/fixtures/modules
lrwxrwxrwx. 1 niek niek 9 Mar 31 15:32 modules/mymodule/spec/fixtures/modules -> ../../../
```

Provision config is:
```
Vagrant.configure("2") do |config|
  config.vm.box = "baremettle/centos-6.5"
  config.ssh.forward_agent = true
  config.vm.synced_folder ".", "/vagrant", type: "rsync", rsync__exclude: [".git/","fixtures/"]

  config.vm.provision :puppet do |puppet|
    puppet.module_path = "modules"
    puppet.manifests_path = "manifests"
    puppet.manifest_file  = "vagrant.pp"
    puppet.options = "--hiera_config /vagrant/hiera.yaml"
  end

  config.vm.define "vm-01" do |vm01_config|
    vm01_config.vm.hostname = 'vm-01.vagrant.lan'
    vm01_config.vm.network :private_network, :ip => "192.168.60.2"
  end
end
```
Running vagrant up results in an rsync when syncing the modules dir:
```
==> vm-01: Rsyncing folder: /home/niek/git/mymodule/modules/ => /tmp/vagrant-puppet-4/modules-0
There was an error when attempting to rsync a synced folder.                                                                                                 
Please inspect the error message below for more info.                                                                                                        

Host path: /home/niek/git/mymodule/modules/                                                                                   
Guest path: /tmp/vagrant-puppet-4/modules-0                                                                                                                  
Command: rsync --verbose --archive --delete -z --copy-links --no-owner --no-group --rsync-path sudo rsync -e ssh -p 22 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -i '/home/niek/.vagrant.d/insecure_private_key' --exclude .vagrant/ /home/niek/git/mymodule/modules/ vagrant@192.168.121.83:/tmp/vagrant-puppet-4/modules-0                                                              
rsync: readlink_stat("/home/niek/git/mymodule/modules/mymodule/spec/fixtures/modules/mymodule..snip..") failed: Too many levels of symbolic links (40)
```

Unfortunately the config.vm.synced_folder options aren't used for the rsync of manifests and modules directory.

**Comment by choffee** [comment]
Just got the same thing. 

By default it looks like the rsync providers adds a --copy-links. Seems to be an odd choice for a default that will cause more problems than it fixes.

Can we get that removed from the rsync module?

**Comment by choffee** [comment]
I have a patch that allows you to set the default args for the filetype so now you can do

config.vm.provision "puppet" do |puppet|
    puppet.manifests_path = "testing"
    puppet.manifest_file  = "vagrant_manifest.pp"
    puppet.module_path = "/my/module/path"
    puppet.synced_folder_args = ['-a', '--delete', '--exclude=fixtures']
    }
  end
Note the new synced_folder_args.

https://github.com/choffee/vagrant/commit/98780028d11f23035a3ae741ee1a06a0bc195686

It works for me but I have yet to dig into the vagrant testing. Does this sound sensible should I create a pull request?

**Comment by sethvargo** [comment]
Fixed in #5359. Thanks!
